File: internal/model/report.go
package model

// ReportType defines the frequency of a usage report schedule.
type ReportType string

const (
	ReportTypeDaily   ReportType = "daily"
	ReportTypeWeekly  ReportType = "weekly"
	ReportTypeMonthly ReportType = "monthly"
)

// ReportMetric defines which data points can be included in a usage report.
type ReportMetric string

const (
	ReportMetricOverview      ReportMetric = "overview"       // 总览：请求数、token、成本、成功率
	ReportMetricTopModels     ReportMetric = "top_models"     // Top 5 模型用量
	ReportMetricTopChannels   ReportMetric = "top_channels"   // Top 5 渠道用量
	ReportMetricTopAPIKeys    ReportMetric = "top_apikeys"    // Top 5 API Key 用量
	ReportMetricCostBreakdown ReportMetric = "cost_breakdown" // 成本明细
	ReportMetricErrorAnalysis ReportMetric = "error_analysis" // 错误分析
	ReportMetricDailyTrend    ReportMetric = "daily_trend"    // 每日趋势（仅周报/月报）
)

// AllReportMetrics returns all valid report metric values.
func AllReportMetrics() []ReportMetric {
	return []ReportMetric{
		ReportMetricOverview,
		ReportMetricTopModels,
		ReportMetricTopChannels,
		ReportMetricTopAPIKeys,
		ReportMetricCostBreakdown,
		ReportMetricErrorAnalysis,
		ReportMetricDailyTrend,
	}
}

// DefaultReportMetrics returns the default metric set for new schedules.
func DefaultReportMetrics() string {
	return `["overview","top_models","top_channels","cost_breakdown"]`
}

// ReportSchedule defines a recurring usage report configuration.
type ReportSchedule struct {
	ID             int        `json:"id" gorm:"primaryKey"`
	Name           string     `json:"name" gorm:"not null"`
	Enabled        bool       `json:"enabled" gorm:"default:true"`
	Type           ReportType `json:"type" gorm:"not null;default:'daily'"`
	Metrics        string     `json:"metrics" gorm:"not null;default:'[\"overview\",\"top_models\",\"top_channels\",\"cost_breakdown\"]'"` // JSON array of ReportMetric
	NotifChannelID int        `json:"notif_channel_id"`
	SendHour       int        `json:"send_hour" gorm:"default:8"`         // 0-23, hour of day to send (in stats timezone)
	SendDayOfWeek  int        `json:"send_day_of_week" gorm:"default:1"`  // 0=Sun..6=Sat, for weekly reports
	SendDayOfMonth int        `json:"send_day_of_month" gorm:"default:1"` // 1-28, for monthly reports
	LastSentAt     int64      `json:"last_sent_at"`                       // unix ms
}

// ReportHistory records a single sent report.
type ReportHistory struct {
	ID           int64      `json:"id" gorm:"primaryKey"`
	ScheduleID   int        `json:"schedule_id" gorm:"index"`
	ScheduleName string     `json:"schedule_name"`
	Type         ReportType `json:"type"`
	Title        string     `json:"title"`
	Content      string     `json:"content"`     // formatted report text
	SendStatus   string     `json:"send_status"` // sent / failed / skipped
	SendDetail   string     `json:"send_detail"` // channel name or error message
	SentAt       int64      `json:"sent_at" gorm:"autoCreateTime:milli"`
}

File: web/src/api/endpoints/report.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '../client';
import { REFETCH_INTERVAL_CONFIG } from '../constants';
import { logger } from '@/lib/logger';

export type ReportType = 'daily' | 'weekly' | 'monthly';
export type ReportMetric =
    | 'overview'
    | 'top_models'
    | 'top_channels'
    | 'top_apikeys'
    | 'cost_breakdown'
    | 'error_analysis'
    | 'daily_trend';

export interface ReportSchedule {
    id: number;
    name: string;
    enabled: boolean;
    type: ReportType;
    notif_channel_id: number;
    metrics: string; // JSON array of metric names
    send_hour: number;
    send_day_of_week: number; // 0-6 for weekly reports
    send_day_of_month: number; // 1-28 for monthly reports
    last_sent_at?: number;
}

export interface ReportHistory {
    id: number;
    schedule_id: number;
    schedule_name: string;
    type: ReportType;
    title: string;
    content: string;
    send_status: 'sent' | 'failed' | 'skipped';
    send_detail: string;
    sent_at: number;
}

export const REPORT_TYPES = [
    { value: 'daily', label: '每日报告' },
    { value: 'weekly', label: '每周报告' },
    { value: 'monthly', label: '每月报告' },
] as const;

export const REPORT_METRICS = [
    { value: 'overview', label: '总览' },
    { value: 'top_models', label: '热门模型' },
    { value: 'top_channels', label: '热门渠道' },
    { value: 'top_apikeys', label: '热门 API Key' },
    { value: 'cost_breakdown', label: '成本明细' },
    { value: 'error_analysis', label: '错误分析' },
    { value: 'daily_trend', label: '每日趋势' },
] as const;

export const DAYS_OF_WEEK = [
    { value: 0, label: '周日' },
    { value: 1, label: '周一' },
    { value: 2, label: '周二' },
    { value: 3, label: '周三' },
    { value: 4, label: '周四' },
    { value: 5, label: '周五' },
    { value: 6, label: '周六' },
] as const;

export function parseReportMetrics(metrics: string): ReportMetric[] {
    try {
        const parsed = JSON.parse(metrics);
        return Array.isArray(parsed) ? parsed : [];
    } catch {
        return [];
    }
}

export function formatReportSendTime(sendHour: number): string {
    const hour = Number.isFinite(sendHour) ? Math.max(0, Math.min(23, sendHour)) : 0;
    return `${String(hour).padStart(2, '0')}:00`;
}

export function parseReportSendHour(sendTime: string): number {
    const hour = Number.parseInt(sendTime.split(':')[0] || '0', 10);
    if (!Number.isFinite(hour)) return 0;
    return Math.max(0, Math.min(23, hour));
}

export function useReportScheduleList() {
    return useQuery({
        queryKey: ['reports', 'schedules'],
        queryFn: async () => apiClient.get<ReportSchedule[]>('/api/v1/report/schedule/list'),
        refetchInterval: REFETCH_INTERVAL_CONFIG,
    });
}

export function useCreateReportSchedule() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (data: Partial<ReportSchedule>) => {
            return apiClient.post<ReportSchedule>('/api/v1/report/schedule/create', data);
        },
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['reports', 'schedules'] });
        },
        onError: (error) => logger.error('Create report schedule failed:', error),
    });
}

export function useUpdateReportSchedule() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (data: ReportSchedule) => {
            return apiClient.post<null>('/api/v1/report/schedule/update', data);
        },
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['reports', 'schedules'] });
        },
        onError: (error) => logger.error('Update report schedule failed:', error),
    });
}

export function useDeleteReportSchedule() {
    const queryClient = useQueryClient();
    return useMutation({
        mutationFn: async (id: number) => {
            return apiClient.delete<null>(`/api/v1/report/schedule/delete/${id}`);
        },
        onSuccess: () => {
            queryClient.invalidateQueries({ queryKey: ['reports', 'schedules'] });
        },
        onError: (error) => logger.error('Delete report schedule failed:', error),
    });
}

export function useTestReportSchedule() {
    return useMutation({
        mutationFn: async (id: number) => {
            return apiClient.post<null>(`/api/v1/report/schedule/test`, { id });
        },
        onError: (error) => logger.error('Test report schedule failed:', error),
    });
}

export function useReportHistory(limit: number = 50) {
    return useQuery({
        queryKey: ['reports', 'history', limit],
        queryFn: async () => apiClient.get<ReportHistory[]>(`/api/v1/report/history/list?limit=${limit}`),
        refetchInterval: REFETCH_INTERVAL_CONFIG,
    });
}
